packages/backend/src/services/PersonalAccessTokenService.ts

```diff
@@ -58,4 +60,42 @@ export class PersonalAccessTokenService extends BaseService {
     ): Promise<PersonalAccessToken[]> {
         return this.personalAccessTokenModel.getAllForUser(user.userId);
     }
+
+    async rotatePersonalAccessToken(
+        user: Pick<SessionUser, 'userUuid'>,
+        personalAccessTokenUuid: string,
+        data: { expiresAt: Date },
+    ): Promise<PersonalAccessTokenWithToken> {
+        if (data.expiresAt.getTime() < Date.now()) {
+            throw new ParameterError('Expire time must be in the future');
+        }
+
+        const existingToken = await this.personalAccessTokenModel.getUserToken({
+            userUuid: user.userUuid,
+            tokenUuid: personalAccessTokenUuid,
+        });
+
+        if (!existingToken.expiresAt) {
+            throw new ParameterError(
+                'Token with no expiration date cannot be rotated',
+            );
+        }
+
+        if (
+            existingToken.rotatedAt &&
+            existingToken.rotatedAt.getTime() > Date.now() - 3600000
+        ) {
+            throw new ParameterError('Token can only be rotated once per hour');
+        }
+
+        const newToken = await this.personalAccessTokenModel.rotate({
+            personalAccessTokenUuid,
+            expiresAt: data.expiresAt,
+        });
+        this.analytics.track({
+            userId: user.userUuid,
+            event: 'personal_access_token.rotated',
```

Comment:
no properties ?

Reply:
None of the personal access tokens has properties. Didn't want to make the PR bigger.